File: 0232-Implement Queue using Stacks/0232-Implement Queue using Stacks.cpp
class MyQueue {
public:
    /** Initialize your data structure here. */
    MyQueue() {
        
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        S1.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        int result = peek();
        S2.pop();
        return result;
    }
    
    /** Get the front element. */
    int peek() {
        if (S2.empty())
        {
            while (!S1.empty())
            {
                S2.push(S1.top());
                S1.pop();
            }
        }
        return S2.top();
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return S1.empty() && S2.empty();
    }

private:
    stack<int> S1, S2;
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * bool param_4 = obj.empty();
 */
 

File: 0201-0300/0232-Implement Queue using Stacks/0232-Implement Queue using Stacks.py
class MyQueue:
    
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.St1 = []
        self.St2 = []

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self.St1.append(x)

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        result = self.peek()
        self.St2.pop()
        return result

    def peek(self) -> int:
        """
        Get the front element.
        """
        if not self.St2:
            while self.St1:
                self.St2.append(self.St1.pop())
                
        return self.St2[-1]

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return len(self.St1) == 0 and len(self.St2) == 0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
